### DISCUSSION TITLE
[Task]: To  find  some  codes

### DISCUSSION THREAD

**Discussion by Yzp1018** [open-edge-platform/anomalib#1574] [Q&A] [answered]
### What is the motivation for this task?

Thanks for the useful tools!

### Describe the solution you'd like

How can I find the codes that calculate the global anomaly scores for a special model when the classification task.Thank you.

**Comment by samet-akcay** [comment]
Hi @Yzp1018, can you elaborate? For example, which model do you want to have a look at?

**Reply by Yzp1018** [reply]
Thank you.
For patchcore,I can find it. I want to know the DRAEM model.I find that your work is different from the original DRAEM model in some details,such as loss ,data augumenter.
So i want to know the difference of the calculation of global anomaly scores.
In the original code,it use the avgpool,            
out_mask_averaged = torch.nn.functional.avg_pool2d(out_mask_sm[:, 1:, :, :], 21, stride=1,
                                                               padding=21 // 2).cpu().detach().numpy()
and then np.max (out_mask_averaged)  .
But in your work ,I cannot find how to calculate it.
Thank you very much.

**Reply by blaz-r** [reply, marked as answer]
If anomaly score isn't separately calculated as part of the model, it is obtained as maximum value of the anomaly map here:

https://github.com/openvinotoolkit/anomalib/blob/df50c0ba150cfefdfe3fd8c745d232b720a281d5/src/anomalib/models/components/base/anomaly_module.py#L186-L193

Regarding the Draem, that is a nice little catch. Draem does in fact contain mean filtering (loccal avgpool) before the max is taken in original implementation. Anomalib version doesn't do that but just takes the max without mean filtering.

**Reply by samet-akcay** [reply]
@djdameln

**Comment by samet-akcay** [comment]
It usually depends on the model, but global anomaly scores are handled by `pred_score` key in the `output` dictionary that is returned by the model. For example, here is the lightning model for Patchcore

https://github.com/openvinotoolkit/anomalib/blob/5f25510f08dab21bc9419d9c0a64be95728e358e/src/anomalib/models/patchcore/lightning_model.py#L115

where the torch model returns `pred_score` and assigns the scores to the batch. To have a closer look at how this `pred_score` is computed, you could refer to the following:

https://github.com/openvinotoolkit/anomalib/blob/5f25510f08dab21bc9419d9c0a64be95728e358e/src/anomalib/models/patchcore/torch_model.py#L82

**Comment by Yzp1018** [comment]
Thank you so much! I find the difference of the training results between the original implementation and your anomalib version.But I do not know why.Now I find the reason.Thanks again!
